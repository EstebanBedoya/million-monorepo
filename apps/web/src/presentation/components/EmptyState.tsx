'use client';

interface EmptyStateProps {
  title?: string;
  message?: string;
  onReset?: () => void;
}

export function EmptyState({ 
  title = "No Properties Found",
  message = "We couldn't find any properties matching your criteria. Try adjusting your filters.",
  onReset 
}: EmptyStateProps) {
  return (
    <div className="flex flex-col items-center justify-center py-16 px-4 text-center" role="status">
      {/* Icon */}
      <div className="mb-6 p-6 rounded-full bg-secondary/10">
        <svg
          className="w-16 h-16 text-secondary"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          aria-hidden="true"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={1.5}
            d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
          />
        </svg>
      </div>

      {/* Content */}
      <h3 className="text-2xl font-semibold text-foreground mb-3">
        {title}
      </h3>
      <p className="text-secondary max-w-md mb-6">
        {message}
      </p>

      {/* Reset button */}
      {onReset && (
        <button
          onClick={onReset}
          className="btn-primary"
          aria-label="Clear all filters"
        >
          Clear All Filters
        </button>
      )}
    </div>
  );
}

