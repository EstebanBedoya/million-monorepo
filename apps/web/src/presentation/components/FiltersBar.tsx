'use client';

import { useState, useEffect, useCallback } from 'react';

export interface FilterValues {
  search: string;
  minPrice: number;
  maxPrice: number;
  propertyType?: string;
}

interface FiltersBarProps {
  onFilterChange: (filters: FilterValues) => void;
  defaultFilters?: Partial<FilterValues>;
}

// Debounce hook
function useDebounce<T>(value: T, delay: number): T {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}

export function FiltersBar({ onFilterChange, defaultFilters }: FiltersBarProps) {
  const [search, setSearch] = useState(defaultFilters?.search || '');
  const [minPrice, setMinPrice] = useState(defaultFilters?.minPrice || 0);
  const [maxPrice, setMaxPrice] = useState(defaultFilters?.maxPrice || 5000000);
  const [propertyType, setPropertyType] = useState(defaultFilters?.propertyType || '');
  const [isExpanded, setIsExpanded] = useState(false);

  // Debounce search input (400ms)
  const debouncedSearch = useDebounce(search, 400);

  // Trigger filter change when debounced values change
  useEffect(() => {
    onFilterChange({
      search: debouncedSearch,
      minPrice,
      maxPrice,
      propertyType,
    });
  }, [debouncedSearch, minPrice, maxPrice, propertyType, onFilterChange]);

  const handleClearFilters = useCallback(() => {
    setSearch('');
    setMinPrice(0);
    setMaxPrice(5000000);
    setPropertyType('');
  }, []);

  const formatPrice = (price: number) => {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(price);
  };

  const hasActiveFilters = search || minPrice > 0 || maxPrice < 5000000 || propertyType;

  return (
    <div className="bg-card border border-border rounded-lg p-6 shadow-sm mb-8">
      {/* Mobile toggle button */}
      <div className="lg:hidden mb-4">
        <button
          onClick={() => setIsExpanded(!isExpanded)}
          className="w-full flex items-center justify-between px-4 py-3 bg-secondary/5 rounded-md"
          aria-expanded={isExpanded}
          aria-controls="filters-content"
        >
          <span className="font-medium text-foreground">Filters</span>
          <svg
            className={`w-5 h-5 text-foreground transition-transform ${isExpanded ? 'rotate-180' : ''}`}
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
          </svg>
        </button>
      </div>

      {/* Filters content */}
      <div
        id="filters-content"
        className={`${isExpanded ? 'block' : 'hidden'} lg:block space-y-6`}
      >
        {/* Search field */}
        <div className="grid grid-cols-1 lg:grid-cols-3 gap-4">
          <div className="lg:col-span-2">
            <label htmlFor="search" className="block text-sm font-medium text-foreground mb-2">
              Search by name or address
            </label>
            <div className="relative">
              <input
                id="search"
                type="text"
                value={search}
                onChange={(e) => setSearch(e.target.value)}
                placeholder="e.g., Modern Villa, Los Angeles..."
                className="input-field pl-10"
                aria-label="Search properties by name or address"
              />
              <svg
                className="absolute left-3 top-1/2 -translate-y-1/2 w-5 h-5 text-secondary pointer-events-none"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                />
              </svg>
            </div>
          </div>

          {/* Property Type */}
          <div>
            <label htmlFor="property-type" className="block text-sm font-medium text-foreground mb-2">
              Property Type
            </label>
            <select
              id="property-type"
              value={propertyType}
              onChange={(e) => setPropertyType(e.target.value)}
              className="input-field"
              aria-label="Filter by property type"
            >
              <option value="">All Types</option>
              <option value="House">House</option>
              <option value="Apartment">Apartment</option>
              <option value="Villa">Villa</option>
              <option value="Condo">Condo</option>
              <option value="Townhouse">Townhouse</option>
              <option value="Studio">Studio</option>
            </select>
          </div>
        </div>

        {/* Price range */}
        <div>
          <div className="flex items-center justify-between mb-3">
            <label className="block text-sm font-medium text-foreground">
              Price Range
            </label>
            <span className="text-sm text-accent font-medium">
              {formatPrice(minPrice)} - {formatPrice(maxPrice)}
            </span>
          </div>

          <div className="space-y-4">
            {/* Min Price Slider */}
            <div>
              <label htmlFor="min-price" className="text-xs text-secondary mb-1 block">
                Minimum Price
              </label>
              <input
                id="min-price"
                type="range"
                min="0"
                max="5000000"
                step="50000"
                value={minPrice}
                onChange={(e) => setMinPrice(Number(e.target.value))}
                className="w-full h-2 bg-secondary/20 rounded-lg appearance-none cursor-pointer 
                           accent-accent focus:outline-none focus:ring-2 focus:ring-accent focus:ring-offset-2"
                aria-label="Minimum price"
                aria-valuemin={0}
                aria-valuemax={5000000}
                aria-valuenow={minPrice}
              />
            </div>

            {/* Max Price Slider */}
            <div>
              <label htmlFor="max-price" className="text-xs text-secondary mb-1 block">
                Maximum Price
              </label>
              <input
                id="max-price"
                type="range"
                min="0"
                max="5000000"
                step="50000"
                value={maxPrice}
                onChange={(e) => setMaxPrice(Number(e.target.value))}
                className="w-full h-2 bg-secondary/20 rounded-lg appearance-none cursor-pointer 
                           accent-accent focus:outline-none focus:ring-2 focus:ring-accent focus:ring-offset-2"
                aria-label="Maximum price"
                aria-valuemin={0}
                aria-valuemax={5000000}
                aria-valuenow={maxPrice}
              />
            </div>
          </div>
        </div>

        {/* Clear filters button */}
        {hasActiveFilters && (
          <div className="pt-4 border-t border-border">
            <button
              onClick={handleClearFilters}
              className="w-full lg:w-auto btn-secondary flex items-center justify-center gap-2"
              aria-label="Clear all filters"
            >
              <svg
                className="w-4 h-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M6 18L18 6M6 6l12 12"
                />
              </svg>
              Clear All Filters
            </button>
          </div>
        )}
      </div>
    </div>
  );
}

