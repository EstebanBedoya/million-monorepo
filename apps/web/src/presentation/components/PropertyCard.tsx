'use client';

import { MockPropertyType } from '../../domain/schemas/property.schema';
import Image from 'next/image';

interface PropertyCardProps {
  property: MockPropertyType;
  onViewDetails?: (property: MockPropertyType) => void;
}

export function PropertyCard({ property, onViewDetails }: PropertyCardProps) {
  // Format price with commas
  const formatPrice = (price: number) => {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(price);
  };

  return (
    <article className="card-elevated overflow-hidden group cursor-pointer" role="article">
      {/* Image container */}
      <div className="relative h-64 overflow-hidden bg-secondary/10">
        <Image
          src={property.image || '/placeholder-property.jpg'}
          alt={property.name}
          fill
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          className="object-cover transition-transform duration-500 group-hover:scale-110"
          onError={(e) => {
            // Fallback to placeholder on error
            const target = e.target as HTMLImageElement;
            target.src = '/placeholder-property.jpg';
          }}
        />
        
        {/* Property type badge */}
        {property.propertyType && (
          <div className="absolute top-4 left-4 bg-card/95 backdrop-blur-sm px-3 py-1.5 rounded-md shadow-sm">
            <span className="text-xs font-medium text-foreground">
              {property.propertyType}
            </span>
          </div>
        )}
      </div>

      {/* Content */}
      <div className="p-6">
        {/* Price */}
        <div className="mb-3">
          <p className="text-3xl font-semibold text-accent">
            {formatPrice(property.price)}
          </p>
        </div>

        {/* Title */}
        <h3 className="text-xl font-semibold text-foreground mb-2 line-clamp-1 group-hover:text-accent transition-colors">
          {property.name}
        </h3>

        {/* Location */}
        <div className="flex items-start gap-2 mb-4">
          <svg 
            className="w-5 h-5 text-secondary flex-shrink-0 mt-0.5" 
            fill="none" 
            stroke="currentColor" 
            viewBox="0 0 24 24"
            aria-hidden="true"
          >
            <path 
              strokeLinecap="round" 
              strokeLinejoin="round" 
              strokeWidth={2} 
              d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z" 
            />
            <path 
              strokeLinecap="round" 
              strokeLinejoin="round" 
              strokeWidth={2} 
              d="M15 11a3 3 0 11-6 0 3 3 0 016 0z" 
            />
          </svg>
          <div className="flex-1 min-w-0">
            <p className="text-sm text-secondary line-clamp-2">
              {property.address}
            </p>
            <p className="text-sm font-medium text-foreground mt-0.5">
              {property.city}
            </p>
          </div>
        </div>

        {/* Property details */}
        {(property.bedrooms || property.bathrooms || property.area) && (
          <div className="flex items-center gap-4 pt-4 border-t border-border">
            {property.bedrooms && (
              <div className="flex items-center gap-1.5" title={`${property.bedrooms} Bedrooms`}>
                <svg 
                  className="w-4 h-4 text-secondary" 
                  fill="none" 
                  stroke="currentColor" 
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                >
                  <path 
                    strokeLinecap="round" 
                    strokeLinejoin="round" 
                    strokeWidth={2} 
                    d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6" 
                  />
                </svg>
                <span className="text-sm text-foreground font-medium">
                  {property.bedrooms} <span className="text-secondary font-normal">bd</span>
                </span>
              </div>
            )}
            
            {property.bathrooms && (
              <div className="flex items-center gap-1.5" title={`${property.bathrooms} Bathrooms`}>
                <svg 
                  className="w-4 h-4 text-secondary" 
                  fill="none" 
                  stroke="currentColor" 
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                >
                  <path 
                    strokeLinecap="round" 
                    strokeLinejoin="round" 
                    strokeWidth={2} 
                    d="M8 14v3m4-3v3m4-3v3M3 21h18M3 10h18M3 7l9-4 9 4M4 10h16v11H4V10z" 
                  />
                </svg>
                <span className="text-sm text-foreground font-medium">
                  {property.bathrooms} <span className="text-secondary font-normal">ba</span>
                </span>
              </div>
            )}
            
            {property.area && (
              <div className="flex items-center gap-1.5" title={`${property.area} ${property.areaUnit || 'm²'}`}>
                <svg 
                  className="w-4 h-4 text-secondary" 
                  fill="none" 
                  stroke="currentColor" 
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                >
                  <path 
                    strokeLinecap="round" 
                    strokeLinejoin="round" 
                    strokeWidth={2} 
                    d="M4 8V4m0 0h4M4 4l5 5m11-1V4m0 0h-4m4 0l-5 5M4 16v4m0 0h4m-4 0l5-5m11 5l-5-5m5 5v-4m0 4h-4" 
                  />
                </svg>
                <span className="text-sm text-foreground font-medium">
                  {property.area} <span className="text-secondary font-normal">{property.areaUnit || 'm²'}</span>
                </span>
              </div>
            )}
          </div>
        )}

        {/* View Details Button (optional) */}
        {onViewDetails && (
          <button
            onClick={() => onViewDetails(property)}
            className="mt-4 w-full btn-primary"
            aria-label={`View details for ${property.name}`}
          >
            View Details
          </button>
        )}
      </div>
    </article>
  );
}
