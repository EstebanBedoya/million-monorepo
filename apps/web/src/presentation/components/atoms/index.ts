export { Button } from './Button';
export { Input } from './Input';
export { Label } from './Label';
export { Icon } from './Icon';
export { Badge } from './Badge';
export { Price } from './Price';
export { Image } from './Image';
export { Select } from './Select';
export { Range } from './Range';
export { Skeleton } from './Skeleton';
export { Spinner } from './Spinner';
export { ConfirmDialog } from './ConfirmDialog';
