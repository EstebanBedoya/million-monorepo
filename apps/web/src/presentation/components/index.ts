// Atomic Design Components Export
export * from './atoms';
export * from './molecules';
export * from './organisms';

// Utility components
export { PropertyList } from './PropertyList';

// Re-export FilterValues type from organisms
export type { FilterValues } from './organisms/FiltersBar';
