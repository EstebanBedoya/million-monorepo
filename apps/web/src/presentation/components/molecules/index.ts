export { SearchInput } from './SearchInput';
export { PriceRange } from './PriceRange';
export { PriceSuggestions } from './PriceSuggestions';
export { PropertyTypeSelect } from './PropertyTypeSelect';
export { PropertyDetails } from './PropertyDetails';
export { LocationInfo } from './LocationInfo';
export { ClearFiltersButton } from './ClearFiltersButton';
export { PageButton } from './PageButton';
export { PageNumberButton } from './PageNumberButton';
export { EmptyState } from './EmptyState';
export { PropertyCardSkeleton, PropertyListSkeleton } from './PropertyCardSkeleton';
export { PaginationSkeleton } from './PaginationSkeleton';
export { ThemeToggle } from './ThemeToggle';
export { AsyncBoundary } from './AsyncBoundary';
export { ErrorState } from './ErrorState';
export { LoadingState } from './LoadingState';
export { Breadcrumbs } from './Breadcrumbs';
