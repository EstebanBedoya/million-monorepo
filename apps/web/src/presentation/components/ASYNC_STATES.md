# Estados de Carga, Error y Vacío

Este documento describe la implementación de estados de carga, error y vacío en la aplicación.

## Componentes Principales

### AsyncBoundary

Componente principal que maneja todos los estados de forma unificada.

```tsx
import { AsyncBoundary } from '../components/molecules/AsyncBoundary';

<AsyncBoundary
  loading={loading}
  error={error}
  isEmpty={!loading && !error && items.length === 0}
  onRetry={handleRetry}
  onClearFilters={handleClearFilters}
  emptyTitle="No Items Found"
  emptyMessage="Try adjusting your filters."
>
  <YourContent />
</AsyncBoundary>
```

### ErrorState

Componente específico para mostrar errores con opción de retry.

```tsx
import { ErrorState } from '../components/molecules/ErrorState';

<ErrorState
  error={error}
  onRetry={handleRetry}
/>
```

### LoadingState

Componente para estados de carga con spinner.

```tsx
import { LoadingState } from '../components/molecules/LoadingState';

<LoadingState
  message="Loading properties..."
  showSpinner={true}
/>
```

## Servicio de Propiedades

### PropertyServiceClient

Servicio genérico para manejar peticiones de propiedades con simulación de errores.

```tsx
import { PropertyServiceClient } from '../application/services/PropertyServiceClient';

const service = new PropertyServiceClient({
  baseUrl: '/api/properties',
  simulateErrors: true,
  errorRate: 0.1,
  errorTypes: ['server', 'timeout', 'network']
});

// Obtener propiedades
const result = await service.getProperties(filters, pagination);

// Configurar simulación de errores
service.setErrorSimulation(true, 0.2);
service.setErrorTypes(['timeout']);
```

### Hook usePropertyService

Hook personalizado que encapsula la lógica del servicio.

```tsx
import { usePropertyService } from '../hooks/usePropertyService';

const {
  properties,
  pagination,
  loading,
  error,
  fetchProperties,
  retry,
  clearError,
  setErrorSimulation
} = usePropertyService({
  baseUrl: '/api/properties',
  simulateErrors: process.env.NODE_ENV === 'development',
  errorRate: 0.1
});
```

## Panel de Simulación de Errores

### ErrorSimulationPanel

Panel de desarrollo para controlar la simulación de errores.

```tsx
import { ErrorSimulationPanel } from '../components/molecules/ErrorSimulationPanel';

<ErrorSimulationPanel
  onSetErrorSimulation={setErrorSimulation}
  onSetErrorTypes={setErrorTypes}
/>
```

**Nota**: Solo aparece en modo desarrollo.

## Tipos de Errores Soportados

1. **Server Error (500)**: Error del servidor
2. **Timeout**: Timeout de la petición
3. **Network**: Error de conexión

## Estados de la Aplicación

### 1. Carga (Loading)
- Muestra skeletons o spinners
- Indica que la petición está en progreso
- Bloquea interacciones del usuario

### 2. Error
- Muestra mensaje de error claro
- Incluye botón de retry
- Diferencia entre tipos de error
- Muestra detalles técnicos en desarrollo

### 3. Vacío (Empty)
- Muestra cuando no hay resultados
- Incluye CTA para limpiar filtros
- Mensaje explicativo

### 4. Éxito (Success)
- Muestra el contenido normal
- Datos cargados correctamente

## Pruebas

Las pruebas cubren todos los estados:

```bash
# Ejecutar pruebas de componentes
npm test AsyncBoundary.test.tsx
npm test ErrorState.test.tsx

# Ejecutar pruebas de servicios
npm test PropertyServiceClient.test.ts
npm test usePropertyService.test.tsx
```

## Criterios de Aceptación ✅

- [x] **Carga**: Skeletons y/o spinner semántico
- [x] **Error**: Mensaje con retry
- [x] **Vacío**: Mensaje + CTA para limpiar filtros
- [x] **AsyncBoundary**: Patrón unificado para todos los estados
- [x] **Simulación de errores**: Sin MSW, servicio genérico
- [x] **Pruebas**: Cobertura para casos ok/empty/error

## DoD (Definition of Done) ✅

- [x] Cobertura de tests para casos: ok/empty/error
- [x] Componentes reutilizables
- [x] Servicio genérico fácil de cambiar
- [x] Documentación completa
- [x] Panel de desarrollo para testing
