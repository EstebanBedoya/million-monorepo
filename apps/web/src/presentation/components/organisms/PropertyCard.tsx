'use client';

import { useRouter } from 'next/navigation';
import { Property } from '../../../domain/entities/Property';
import { Image } from '../atoms/Image';
import { Badge } from '../atoms/Badge';
import { Price } from '../atoms/Price';
import { Button } from '../atoms/Button';
import { LocationInfo } from '../molecules/LocationInfo';
import { PropertyDetails } from '../molecules/PropertyDetails';

export interface PropertyCardProps {
  property: Property;
  onViewDetails?: (property: Property) => void;
}

export function PropertyCard({ property, onViewDetails }: PropertyCardProps) {
  const router = useRouter();

  const handleViewDetails = () => {
    if (onViewDetails) {
      onViewDetails(property);
    } else {
      router.push(`/properties/${property.id}`);
    }
  };

  return (
    <article className="card-elevated overflow-hidden group cursor-pointer" role="article">
      <div className="relative h-64 overflow-hidden bg-secondary/10">
        <Image
          src={property.images[0]}
          alt={property.name}
          fill
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          className="object-cover transition-transform duration-500 group-hover:scale-110"
          fallbackSrc="/placeholder-property.jpg"
        />
        
        {property.propertyType && (
          <div className="absolute top-4 left-4">
            <Badge variant="secondary" className="bg-gray-700/50 text-gray-100 backdrop-blur-sm shadow-sm">
              {property.propertyType}
            </Badge>
          </div>
        )}
      </div>

      <div className="p-6">
        <div className="mb-3">
          <Price amount={property.price} size="xl" />
        </div>

        <h3 className="text-xl font-semibold text-foreground mb-2 line-clamp-1 group-hover:text-accent transition-colors">
          {property.name}
        </h3>

        <LocationInfo 
          address={property.location}
          className="mb-4"
        />

        <PropertyDetails
          bedrooms={property.bedrooms}
          bathrooms={property.bathrooms}
          area={property.area}
          areaUnit={property.areaUnit}
        />

        <Button
          onClick={handleViewDetails}
          className="mt-4 w-full"
          aria-label={`View details for ${property.name}`}
        >
          View Details
        </Button>
      </div>
    </article>
  );
}
