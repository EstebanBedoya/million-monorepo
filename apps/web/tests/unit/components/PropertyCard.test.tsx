import { render, screen } from '@testing-library/react';
import { PropertyCard } from '../../../src/presentation/components/PropertyCard';
import { MockPropertyType } from '../../../src/domain/schemas/property.schema';

// Mock Next.js Image component
jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line @next/next/no-img-element, jsx-a11y/alt-text
    return <img {...props} />;
  },
}));

describe('PropertyCard', () => {
  const mockProperty: MockPropertyType = {
    id: 'prop-001',
    idOwner: 'owner-001',
    name: 'Modern Luxury Villa',
    address: '1247 Sunset Boulevard',
    city: 'Los Angeles',
    price: 2800000,
    image: 'https://example.com/image.jpg',
    bedrooms: 5,
    bathrooms: 4,
    area: 84,
    areaUnit: 'm²',
    propertyType: 'Villa',
  };

  it('should render property information correctly', () => {
    render(<PropertyCard property={mockProperty} />);

    expect(screen.getByText('Modern Luxury Villa')).toBeInTheDocument();
    expect(screen.getByText('$2,800,000')).toBeInTheDocument();
    expect(screen.getByText('Los Angeles')).toBeInTheDocument();
    expect(screen.getByText('1247 Sunset Boulevard')).toBeInTheDocument();
  });

  it('should display property details when available', () => {
    render(<PropertyCard property={mockProperty} />);

    // Check for bedrooms, bathrooms, and area with more specific selectors
    expect(screen.getByTitle('5 Bedrooms')).toBeInTheDocument();
    expect(screen.getByTitle('4 Bathrooms')).toBeInTheDocument();
    expect(screen.getByTitle('84 m²')).toBeInTheDocument();
  });

  it('should render without optional fields', () => {
    const minimalProperty: MockPropertyType = {
      id: 'prop-002',
      idOwner: 'owner-002',
      name: 'Simple Apartment',
      address: '123 Main St',
      city: 'New York',
      price: 500000,
      image: 'https://example.com/apt.jpg',
    };

    render(<PropertyCard property={minimalProperty} />);

    expect(screen.getByText('Simple Apartment')).toBeInTheDocument();
    expect(screen.getByText('$500,000')).toBeInTheDocument();
  });

  it('should display property type badge when available', () => {
    render(<PropertyCard property={mockProperty} />);

    expect(screen.getByText('Villa')).toBeInTheDocument();
  });

  it('should call onViewDetails when button is clicked', () => {
    const handleViewDetails = jest.fn();
    render(<PropertyCard property={mockProperty} onViewDetails={handleViewDetails} />);

    const button = screen.getByRole('button', { name: /View details for Modern Luxury Villa/i });
    button.click();

    expect(handleViewDetails).toHaveBeenCalledWith(mockProperty);
  });

  it('should not render View Details button when onViewDetails is not provided', () => {
    render(<PropertyCard property={mockProperty} />);

    const button = screen.queryByRole('button', { name: /View details/i });
    expect(button).not.toBeInTheDocument();
  });

  it('should have proper accessibility attributes', () => {
    render(<PropertyCard property={mockProperty} />);

    const article = screen.getByRole('article');
    expect(article).toBeInTheDocument();

    const image = screen.getByAltText('Modern Luxury Villa');
    expect(image).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<PropertyCard property={mockProperty} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
