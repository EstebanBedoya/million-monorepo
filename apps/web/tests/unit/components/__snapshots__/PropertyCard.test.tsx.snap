// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyCard should match snapshot 1`] = `
<article
  class="card-elevated overflow-hidden group cursor-pointer"
  role="article"
>
  <div
    class="relative h-64 overflow-hidden bg-secondary/10"
  >
    <img
      alt="Modern Luxury Villa"
      class="object-cover transition-transform duration-500 group-hover:scale-110"
      sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
      src="https://example.com/image.jpg"
    />
    <div
      class="absolute top-4 left-4 bg-card/95 backdrop-blur-sm px-3 py-1.5 rounded-md shadow-sm"
    >
      <span
        class="text-xs font-medium text-foreground"
      >
        Villa
      </span>
    </div>
  </div>
  <div
    class="p-6"
  >
    <div
      class="mb-3"
    >
      <p
        class="text-3xl font-semibold text-accent"
      >
        $2,800,000
      </p>
    </div>
    <h3
      class="text-xl font-semibold text-foreground mb-2 line-clamp-1 group-hover:text-accent transition-colors"
    >
      Modern Luxury Villa
    </h3>
    <div
      class="flex items-start gap-2 mb-4"
    >
      <svg
        aria-hidden="true"
        class="w-5 h-5 text-secondary flex-shrink-0 mt-0.5"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
        <path
          d="M15 11a3 3 0 11-6 0 3 3 0 016 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      <div
        class="flex-1 min-w-0"
      >
        <p
          class="text-sm text-secondary line-clamp-2"
        >
          1247 Sunset Boulevard
        </p>
        <p
          class="text-sm font-medium text-foreground mt-0.5"
        >
          Los Angeles
        </p>
      </div>
    </div>
    <div
      class="flex items-center gap-4 pt-4 border-t border-border"
    >
      <div
        class="flex items-center gap-1.5"
        title="5 Bedrooms"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-secondary"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="text-sm text-foreground font-medium"
        >
          5
           
          <span
            class="text-secondary font-normal"
          >
            bd
          </span>
        </span>
      </div>
      <div
        class="flex items-center gap-1.5"
        title="4 Bathrooms"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-secondary"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 14v3m4-3v3m4-3v3M3 21h18M3 10h18M3 7l9-4 9 4M4 10h16v11H4V10z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="text-sm text-foreground font-medium"
        >
          4
           
          <span
            class="text-secondary font-normal"
          >
            ba
          </span>
        </span>
      </div>
      <div
        class="flex items-center gap-1.5"
        title="84 m²"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-secondary"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 8V4m0 0h4M4 4l5 5m11-1V4m0 0h-4m4 0l-5 5M4 16v4m0 0h4m-4 0l5-5m11 5l-5-5m5 5v-4m0 4h-4"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="text-sm text-foreground font-medium"
        >
          84
           
          <span
            class="text-secondary font-normal"
          >
            m²
          </span>
        </span>
      </div>
    </div>
  </div>
</article>
`;
